### PR TITLE
Upgrade to core-js@3.

### PR DESCRIPTION
**Pull request by sgomes**
It's starting to take a toll to still use `core-js@2`, since we're missing important features: `Object.fromEntries`, `Array.flat`, `globalThis`, `URL`, to name a few. We've been adding support for some of these with individual polyfills, but it's time to take the leap and move over to `core-js@3`, to ensure our development environment stays representative of the current state of the JS spec.

This will initially result in some code duplication, as unfortunately we use several packages with a hardcoded dependency on `core-js@2`. Hopefully, over time, we'll be able to replace these packages with more modern ones, or update them ourselves to be compatible with `core-js@3`.

#### Changes proposed in this Pull Request

* Update the `core-js` package to the current version.
* Set the `core-js` version in the babel config.
* Drop `@babel/polyfill` and use `core-js` directly.
* Remove superfluous polyfills (e.g. `@webcomponents/url`).
* Remove superfluous polyfill module replacements in evergreen.
* Modify server entry point to use `boot/polyfills` rather than its own code.
* Minor lint fixes.

#### Testing instructions

This change affects all of Calypso, so please be sure to try using it as much as possible both in newer browsers, which should use the evergreen build (that includes fewer polyfills); and in older browsers (particularly IE11), which should use the fallback build (that includes all polyfills).

Potential areas of interest: anything that does URL processing (e.g. sharing from Reader), and anything that uses `globalThis` (e.g. language switching).

### CODE REVIEW

File: client/boot/polyfills.js [hunk 1]
```diff
@@ -1,24 +1,23 @@
 /**
  * External dependencies
  */
-import '@babel/polyfill';
+import 'core-js/stable';
+import 'regenerator-runtime/runtime';
```
Comment (sgomes) [inline client/boot/polyfills.js]: Note: we need to depend on this directly now, as it used to be included by `@babel/polyfill`, but it's not in `core-js/stable`. See https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md#babelpolyfill

**Comment by sgomes** [comment]
Marking the PR as "In Progress" while I track down some IE11 bugs.

File: apps/notifications/src/standalone/index.js [hunk 2]
```diff
@@ -1,12 +1,21 @@
-import '@babel/polyfill';
+/**
+ * External dependencies
+ */
+import 'boot/polyfills';
```
Comment (jsnajdr) [inline apps/notifications/src/standalone/index.js]: Does this import really resolve correctly for a standalone Notifications app that's not part of Calypso app? These imports shouldn't automatically resolve to `client/`

**Review by jsnajdr — COMMENTED** [review]
The `boot/polyfills` module would deserve its own NPM package (`@automattic/calypso-polyfills`?) 🙂 We'll need to import it in the `wp-desktop` app, too, together with starting to build it with `calypso-build`, and remove custom/duplicate code like parts of https://github.com/Automattic/wp-desktop/pull/733

File: apps/notifications/src/standalone/index.js (hunk 2 above)
Comment (sgomes) [reply apps/notifications/src/standalone/index.js]: You're right, this does not work! I'll go ahead and turn the polyfills into a package, as you suggested 👍

Comment (jsnajdr) [reply apps/notifications/src/standalone/index.js]: A npm package might be more challenging than I originally thought: the env preset needs to transform the `core-js` imports inside it, differently for each build target.

Comment (sgomes) [reply apps/notifications/src/standalone/index.js]: Yes, you're right 😞 I'll use just the bare `core-js` polyfills (+regenerator) here.

**Comment by sgomes** [comment]
`String.split` seems to be behaving differently in this branch, in IE 11.

The following code produces very different output between production and this branch:

```
var x = "By continuing, you agree to our {{tosLink}}Terms of Service{{/tosLink}}.";
x.split( /(\{\{\/?\s*\w+\s*\/?\}\})/g );
```

This generates an array with 5 items (as expected) in production, and 143 (!) in this branch.

There appears to be a bug in `core-js@3.6.1`: zloirock/core-js#741 . Trying update to latest version (3.6.2), released today.

**Comment by sgomes** [comment]
Back on track, after updating to `core-js@3.6.2`; no more obvious errors on IE 11. Placing back under 'Needs Review'.

**Comment by sgomes** [comment]
@jsnajdr This PR appears to be significantly inflating bundle sizes, upwards of 14KB for most entry points. If we are to deploy this, we should probably prioritise getting rid of the remnants of `core-js@2` quickly.

**Comment by jsnajdr** [comment]
> This PR appears to significantly inflate bundle sizes, upwards of 14KB (compressed!)

Yes, that's expected 🙂 Let's remove the offending libraries in the following few days.

**Comment by sgomes** [comment]
After rebasing, all tests are now green.

**Comment by sgomes** [comment]
Rebased again, this time due to #38688.

File: packages/calypso-build/babel/default.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ module.exports = ( api, opts ) => ( {
 		[
 			require.resolve( '@babel/preset-env' ),
 			{
-				corejs: 2,
+				corejs: 3.6,
```
Comment (blowery) [inline packages/calypso-build/babel/default.js]: are there docs that show this is a supported value? The babel docs seem to want to make this `2` or `3`. Having it be a float seems odd, I'd expect a semver string. https://babeljs.io/docs/en/babel-preset-env#corejs

Comment (sgomes) [reply packages/calypso-build/babel/default.js]: I was following https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md#babelpreset-env , but there it appears to need to be a string.

Looking at the code, this *should* work, but given the lack of mention in the official docs, it's probably best to switch to just `3` here.

Done, thanks for the suggestion!

Comment (jsnajdr) [reply packages/calypso-build/babel/default.js]: I had this same question, and also checked this in the Babel code. The logic is to use CoreJS 2 if the value is `2`, and to use CoreJS 3 if the value is anything else, including `3.6` 🙂

**Comment by blowery** [comment]
The latest bundle size reports seem a bit odd. Removed 56k from entrypoints? Maybe the report is just bad?

**Comment by sgomes** [comment]
> The latest bundle size reports seem a bit odd. Removed 56k from entrypoints? Maybe the report is just bad?

Looking at the live branch file sizes, the report appears to be bad. The 14KB bump per entry point is still there.

**Comment by blowery** [comment]
Looks like react-virtualized might be getting core-js@3 support some time soon. https://github.com/bvaughn/react-virtualized/commit/48d3566b66c30047dcfc84efdec86ee55845e790 landed right after the current release which moves up.

**Comment by sgomes** [comment]
> Looks like react-virtualized might be getting core-js@3 support some time soon. [bvaughn/react-virtualized@48d3566](https://github.com/bvaughn/react-virtualized/commit/48d3566b66c30047dcfc84efdec86ee55845e790) landed right after the current release which moves up.

Thanks, @blowery! I had happened to noticed this; it's great to see things moving!

This is still very much suboptimal, however. In doing this, it's still depending on the `core-js` implementation directly, by calling these methods to avoid having to rely on the browser being modern or global polyfills being present. In other words, it's choosing to handle polyfilling for the library itself, rather than handing off that task to the consumer.

The issue is that the method calls and the methods will always be there, even if we're running in an environment which has native support for the functionality they're using. This effectively means that we'd be bundling the polyfill implementation code (even if we're not polyfilling) in the evergreen build. It may be better than bundling the `core-js` 2 implementation, but my guess is not by much.

The real solution here, as I see it, is to bite the bullet and roll out our own `react-virtualized`, based on the current master. We'd ship a package that includes the sources (or something close to the sources; we'd just strip away the flow stuff), and we'd compile the package by whitelisting it in the Calypso build.

I've played around with modifying the `react-virtualized` build system to allow for the above, and it seems like it would work correctly, if we want to go down that route.

**Comment by sgomes** [comment]
Rebased on master to fix conflicts. Squashed everything into a single commit to make it easier.

**Comment by sgomes** [comment]
The E2E failure on IE seems to be unrelated to this PR, and probably an issue with the E2E infrastructure itself, as there are mentions of driver issues in the logs.

I tried both logging in and going through the full sign-up process for a new site in IE11 and couldn't find any issues.

**Comment by blowery** [comment]
> This is still very much suboptimal, however.

Yup, totally agreed

> The real solution here, as I see it, is to bite the bullet and roll out our own react-virtualized, based on the current master. 

Maybe we could offer assistance to maintain it? Or do you think forking is worth it?

**Comment by jsnajdr** [comment]
> The `globalThis` issue should be fixed now, by requiring it explicitly. It's unclear why this is required

After a fun debugging session, I discovered that we need to specify `corejs: 3.6` instead of `corejs: 3` to make this work.

When listing the expanded available polyfills that `core-js/stable` expands to, the env preset looks at the [`core-js-compat/modules-by-versions.json`](https://unpkg.com/browse/core-js-compat@3.5.0/modules-by-versions.json) file. And this file says that `es.global-this` was added only in `core-js@3.3`. When using `corejs: 3`, it's coerced to `3.0.0` and fails to include this polyfill.

**Comment by sgomes** [comment]
> > The real solution here, as I see it, is to bite the bullet and roll out our own react-virtualized, based on the current master.
> 
> Maybe we could offer assistance to maintain it? Or do you think forking is worth it?

I'm just not sure our use-case matches anyone else's exactly. What we want, and probably what a lot of folks want, is for the package to include sources that they can compile, so that polyfills are all taken care of as part of the holistic build process. However, what those sources should look like is different to everyone.

For example, a lot of folks use `flow`, so they may be ok with using the project sources exactly as they are. We don't, so we'd need the "sources" in the package to have `flow` types stripped. We have support for the class properties proposal in Calypso, some other projects might not.

Perhaps we could agree on shipping an "ES20xx" build in the package, but which "xx"? And should it include JSX as-is, or have it transpiled (and if so, using which options?)

The fundamental issue is that there is no universal "source" that can be distributed to every project in such a way that it can always be compiled, so they decided to distribute a "compiled" version, and in doing so tied their library to `core-js` for the features where a simple syntax transform isn't enough. I can't see a generic enough solution to that problem that would be worth considering as part of the core project, but if one exists, I'm more than happy to put time towards implementing it and trying to upstream it.

**Comment by jsnajdr** [comment]
Other stable polyfills we were losing because of `corejs: 3`:
- `JSON.stringify` patched to comply with https://github.com/tc39/proposal-well-formed-stringify
- `String.matchAll` and `String.replaceAll`
- `Promise.allSettled`

**Comment by sgomes** [comment]
> > The `globalThis` issue should be fixed now, by requiring it explicitly. It's unclear why this is required
> 
> After a fun debugging session, I discovered that we need to specify `corejs: 3.6` instead of `corejs: 3` to make this work.
> 
> When listing the expanded available polyfills that `core-js/stable` expands to, the env preset looks at the [`core-js-compat/modules-by-versions.json`](https://unpkg.com/browse/core-js-compat@3.5.0/modules-by-versions.json) file. And this file says that `es.global-this` was added only in `core-js@3.3`. When using `corejs: 3`, it's coerced to `3.0.0` and fails to include this polyfill.

Excellent sleuthing, @jsnajdr 👍 I've switched back to `corejs: 3.6` and removed the workaround.

**Comment by jsnajdr** [comment]
After researching `create-react-app`, it turns out that standalone `@automattic/calypso-polyfills` package is possible and not hard to do. They do something very similar in their `react-app-polyfill` package.

The thing that makes this possible is the fact that when transpiling dependencies, both CRA and Calypso use env preset with the `useBuiltIns: 'entry'` flag and therefore expand `core-js` imports even in `node_modules` dependencies.

The Calypso dependencies are transpiled with the [babel.dependencies.config.js](https://github.com/Automattic/wp-calypso/blob/update/use-core-js-3/babel.dependencies.config.js#L9) config. And it also needs update from `corejs: 2` and `corejs: 3.6`. It's entirely possible that part of the bundle size inflation comes from this omission!

**Comment by jsnajdr** [comment]
> However, what those sources should look like is different to everyone.

The best common denominator at this moment seems to be standard JavaScript in the latest version (at package publish time). That's also how CRA and Calypso transpile their `node_modules` dependencies. Standard JS without transpiling standard features like classes, ESM modules and no syntax extensions like JSX or Flow.

It shouldn't be hard to reach agreement on that. I'd prefer to avoid creating our own fork of `react-virtualized`, especially when we are not adding any features or fixing any serious bugs.

**Comment by sgomes** [comment]
> The best common denominator at this moment seems to be standard JavaScript in the latest version (at package publish time). That's also how CRA and Calypso transpile their `node_modules` dependencies. Standard JS without transpiling standard features like classes, ESM modules and no syntax extensions like JSX or Flow.
> 
> It shouldn't be hard to reach agreement on that. I'd prefer to avoid creating our own fork of `react-virtualized`, especially when we are not adding any features or fixing any serious bugs.

Fair enough, I'll work on a PR and pitch the idea to the project 👍

**Comment by sgomes** [comment]
> After researching `create-react-app`, it turns out that standalone `@automattic/calypso-polyfills` package is possible and not hard to do. They do something very similar in their `react-app-polyfill` package.

Looking at `react-app-polyfill` it does seem very doable! I'll rework the PR to turn it into a package 👍

**Comment by sgomes** [comment]
> The Calypso dependencies are transpiled with the [babel.dependencies.config.js](https://github.com/Automattic/wp-calypso/blob/update/use-core-js-3/babel.dependencies.config.js#L9) config. And it also needs update from `corejs: 2` and `corejs: 3.6`. It's entirely possible that part of the bundle size inflation comes from this omission!

Good catch! Fixed.

**Comment by sgomes** [comment]
@jsnajdr : The polyfills have been package-ified!

File: packages/calypso-polyfills/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+## Usage
+
+In node, simply require the package:
+
+```js
+  require( '@automattic/calypso-polyfills' );
+```
+
+In the client, a similar naked import will include the polyfills (defaulting to the `fallback` set):
```
Comment (jsnajdr) [inline packages/calypso-polyfills/README.md]: Nit: "In a browser" is more precise wording than "In the client"

File: packages/calypso-polyfills/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## Features
+
+`calypso-polyfills` ties into the Calypso build process, to ensure that only the required polyfills for each build target are needed. Through the use of Babel's `preset-env` and `browserslist` configurations, we strip out any polyfills that are supported for all browsers that are supported by that build target.
+
+See the root directory's `package.json` for the list of supported browsers for each build target.
+
+## Usage
+
+In node, simply require the package:
```
Comment (jsnajdr) [inline packages/calypso-polyfills/README.md]: "In Node.js" 🙂

File: packages/calypso-polyfills/index.node.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Internal dependencies
+ */
+require( './node' );
```
Comment (jsnajdr) [inline packages/calypso-polyfills/index.node.js]: The `index.node.js` and `index.web.js` files are a bit redundant. You could directly specify the target files in `package.json`:
```
  "main": "node.js",
  "browser": "browser-fallback.js"
```
Using names `browser-evergreen.js` and `browser-fallback.js` (i.e., with prefix) is also a good idea in my opinion.

**Review by jsnajdr — APPROVED** [review]
This is awesome ❤️ I love how the `core-js` package and other polyfills got pushed down to a leaf `package.json`.

File: packages/calypso-polyfills/README.md (hunk 4 above)
Comment (sgomes) [reply packages/calypso-polyfills/README.md]: Fixed.

File: packages/calypso-polyfills/README.md (hunk 5 above)
Comment (sgomes) [reply packages/calypso-polyfills/README.md]: Fixed.

File: packages/calypso-polyfills/index.node.js (hunk 6 above)
Comment (sgomes) [reply packages/calypso-polyfills/index.node.js]: Good ideas! Done!

File: webpack.config.js [hunk 7]
```diff
@@ -79,6 +79,7 @@ const nodeModulesToTranspile = [
 	'striptags/',
 	'unicode-match-property-ecmascript/',
 	'unicode-match-property-value-ecmascript/',
+	'gridicons/',
```
Comment (jsnajdr) [inline webpack.config.js]: Is the `gridicons/` transpilation related to the CoreJS 3 upgrade?

In any case, you need to also add `@automattic/calypso-polyfills` to the whitelist to make it all work.

Comment (sgomes) [reply webpack.config.js]: Possibly unrelated. This was a fix that was required for production builds to work locally for me, but it doesn't seem to be an issue outside of my machine. I'll remove that and double-check that everything still works fine.

File: webpack.config.js [hunk 8]
```diff
@@ -64,6 +64,7 @@ const nodeModulesToTranspile = [
 	// general form is <package-name>/.
 	// The trailing slash makes sure we're not matching these as prefixes
 	// In some cases we do want prefix style matching (lodash. for lodash.assign)
+	'@automattic/calypso-polyfills',
```
Comment (jsnajdr) [inline webpack.config.js]: There needs to be a trailing slash here. If there isn't one, the code would also match a package name like `@automattic/calypso-polyfills-blah-blah` 🙂

Comment (sgomes) [reply webpack.config.js]: Oops, I missed that, sorry!

**Pull request merged**